File: qmux/src/frame.rs
use bytes::{Buf, BufMut, Bytes, BytesMut};
use web_transport_proto::VarInt;

use crate::{Error, StreamId, Version};

#[derive(Debug, Clone)]
pub struct Stream {
    pub id: StreamId,
    pub data: Bytes,
    pub fin: bool,
}

#[derive(Debug, Clone)]
pub struct ResetStream {
    pub id: StreamId,
    pub code: VarInt,
}

#[derive(Debug, Clone)]
pub struct StopSending {
    pub id: StreamId,
    pub code: VarInt,
}

#[derive(Debug, Clone)]
pub struct ConnectionClose {
    pub code: VarInt,
    pub reason: String,
}

/// QUIC-compatible frames for multiplexed transport
#[derive(Debug)]
pub enum Frame {
    ResetStream(ResetStream),
    StopSending(StopSending),
    ConnectionClose(ConnectionClose),
    Stream(Stream),
}

impl Frame {
    pub fn encode(&self, version: Version) -> Bytes {
        let mut buf = BytesMut::new();

        match version {
            Version::WebTransport => self.encode_wt(&mut buf),
            Version::QMux00 => self.encode_qmux(&mut buf),
        }

        buf.freeze()
    }

    fn encode_wt(&self, buf: &mut BytesMut) {
        match self {
            Frame::Stream(s) => {
                buf.put_u8(if s.fin { 0x09 } else { 0x08 });
                s.id.0.encode(buf);
                buf.put_slice(&s.data);
            }
            Frame::ResetStream(r) => {
                buf.put_u8(0x04);
                r.id.0.encode(buf);
                r.code.encode(buf);
            }
            Frame::StopSending(s) => {
                buf.put_u8(0x05);
                s.id.0.encode(buf);
                s.code.encode(buf);
            }
            Frame::ConnectionClose(c) => {
                buf.put_u8(0x1d);
                c.code.encode(buf);
                buf.put_slice(c.reason.as_bytes());
            }
        }
    }

    fn encode_qmux(&self, buf: &mut BytesMut) {
        match self {
            Frame::Stream(s) => {
                // Always LEN bit (0x02), never OFF bit. Type = 0x0a | fin_bit
                let frame_type = 0x08u64 | 0x02 | if s.fin { 0x01 } else { 0 };
                VarInt::try_from(frame_type).unwrap().encode(buf);
                s.id.0.encode(buf);
                VarInt::try_from(s.data.len() as u64).unwrap().encode(buf);
                buf.put_slice(&s.data);
            }
            Frame::ResetStream(r) => {
                VarInt::try_from(0x04u64).unwrap().encode(buf);
                r.id.0.encode(buf);
                r.code.encode(buf);
                // final_size = 0 (no flow control tracking)
                VarInt::from(0u32).encode(buf);
            }
            Frame::StopSending(s) => {
                VarInt::try_from(0x05u64).unwrap().encode(buf);
                s.id.0.encode(buf);
                s.code.encode(buf);
            }
            Frame::ConnectionClose(c) => {
                // APPLICATION_CLOSE (0x1d)
                VarInt::try_from(0x1du64).unwrap().encode(buf);
                c.code.encode(buf);
                // frame_type = 0 (application close)
                VarInt::from(0u32).encode(buf);
                let reason_bytes = c.reason.as_bytes();
                VarInt::try_from(reason_bytes.len() as u64).unwrap().encode(buf);
                buf.put_slice(reason_bytes);
            }
        }
    }

    #[allow(dead_code)]
    pub fn decode(data: Bytes, version: Version) -> Result<Option<Self>, Error> {
        if data.is_empty() {
            return Err(Error::Short);
        }

        match version {
            Version::WebTransport => Self::decode_wt(data).map(Some),
            Version::QMux00 => Self::decode_qmux(data),
        }
    }

    fn decode_wt(mut data: Bytes) -> Result<Self, Error> {
        let frame_type = data.get_u8();

        match frame_type {
            0x04 => {
                let id = StreamId(VarInt::decode(&mut data)?);
                let code = VarInt::decode(&mut data)?;
                Ok(Frame::ResetStream(ResetStream { id, code }))
            }
            0x05 => {
                let id = StreamId(VarInt::decode(&mut data)?);
                let code = VarInt::decode(&mut data)?;
                Ok(Frame::StopSending(StopSending { id, code }))
            }
            0x08 => {
                let id = StreamId(VarInt::decode(&mut data)?);
                Ok(Frame::Stream(Stream { id, data, fin: false }))
            }
            0x09 => {
                let id = StreamId(VarInt::decode(&mut data)?);
                Ok(Frame::Stream(Stream { id, data, fin: true }))
            }
            0x1d => {
                let code = VarInt::decode(&mut data)?;
                let reason = String::from_utf8_lossy(&data).into_owned();
                Ok(Frame::ConnectionClose(ConnectionClose { code, reason }))
            }
            _ => Err(Error::InvalidFrameType(frame_type as u64)),
        }
    }

    fn decode_qmux(mut data: Bytes) -> Result<Option<Self>, Error> {
        let frame_type = VarInt::decode(&mut data)?.into_inner();

        // STREAM frames: 0x08-0x0f
        if (0x08..=0x0f).contains(&frame_type) {
            let has_off = frame_type & 0x04 != 0;
            let has_len = frame_type & 0x02 != 0;
            let has_fin = frame_type & 0x01 != 0;

            let id = StreamId(VarInt::decode(&mut data)?);

            if has_off {
                let _offset = VarInt::decode(&mut data)?;
            }

            let stream_data = if has_len {
                let len = VarInt::decode(&mut data)?.into_inner() as usize;
                if data.remaining() < len {
                    return Err(Error::Short);
                }
                data.split_to(len)
            } else {
                data.split_to(data.remaining())
            };

            return Ok(Some(Frame::Stream(Stream {
                id,
                data: stream_data,
                fin: has_fin,
            })));
        }

        match frame_type {
            // RESET_STREAM
            0x04 => {
                let id = StreamId(VarInt::decode(&mut data)?);
                let code = VarInt::decode(&mut data)?;
                let _final_size = VarInt::decode(&mut data)?;
                Ok(Some(Frame::ResetStream(ResetStream { id, code })))
            }
            // STOP_SENDING
            0x05 => {
                let id = StreamId(VarInt::decode(&mut data)?);
                let code = VarInt::decode(&mut data)?;
                Ok(Some(Frame::StopSending(StopSending { id, code })))
            }
            // CONNECTION_CLOSE / APPLICATION_CLOSE
            0x1c | 0x1d => {
                let code = VarInt::decode(&mut data)?;
                let _frame_type = VarInt::decode(&mut data)?;
                let reason_len = VarInt::decode(&mut data)?.into_inner() as usize;
                if data.remaining() < reason_len {
                    return Err(Error::Short);
                }
                let reason = String::from_utf8_lossy(&data.split_to(reason_len)).into_owned();
                Ok(Some(Frame::ConnectionClose(ConnectionClose { code, reason })))
            }
            // MAX_DATA
            0x10 => {
                let _max = VarInt::decode(&mut data)?;
                Ok(None)
            }
            // MAX_STREAM_DATA
            0x11 => {
                let _id = VarInt::decode(&mut data)?;
                let _max = VarInt::decode(&mut data)?;
                Ok(None)
            }
            // MAX_STREAMS (bidi/uni)
            0x12 | 0x13 => {
                let _max = VarInt::decode(&mut data)?;
                Ok(None)
            }
            // DATA_BLOCKED
            0x14 => {
                let _limit = VarInt::decode(&mut data)?;
                Ok(None)
            }
            // STREAM_DATA_BLOCKED
            0x15 => {
                let _id = VarInt::decode(&mut data)?;
                let _limit = VarInt::decode(&mut data)?;
                Ok(None)
            }
            // STREAMS_BLOCKED (bidi/uni)
            0x16 | 0x17 => {
                let _limit = VarInt::decode(&mut data)?;
                Ok(None)
            }
            // DATAGRAM / DATAGRAM with length
            0x30 | 0x31 => Ok(None),
            // QX_TRANSPORT_PARAMETERS
            0x3f5153300d0a0d0a => Ok(None),
            // Unknown — ignore
            _ => Ok(None),
        }
    }
}

impl From<Stream> for Frame {
    fn from(stream: Stream) -> Self {
        Frame::Stream(stream)
    }
}

impl From<ResetStream> for Frame {
    fn from(reset: ResetStream) -> Self {
        Frame::ResetStream(reset)
    }
}

impl From<StopSending> for Frame {
    fn from(stop: StopSending) -> Self {
        Frame::StopSending(stop)
    }
}

impl From<ConnectionClose> for Frame {
    fn from(close: ConnectionClose) -> Self {
        Frame::ConnectionClose(close)
    }
}

File: qmux/src/frame.ts
import * as Stream from "./stream.ts";
import { VarInt } from "./varint.ts";

export type Version = "webtransport" | "qmux-00";

/** Maximum size of a single QMux frame on the wire. */
export const MAX_FRAME_SIZE = 16384;

/** Maximum payload per STREAM frame, accounting for frame overhead (24 bytes). */
export const MAX_FRAME_PAYLOAD = MAX_FRAME_SIZE - 24;

export interface Data {
	type: "stream";
	id: Stream.Id;
	data: Uint8Array;
	fin: boolean;
}

export interface ResetStream {
	type: "reset_stream";
	id: Stream.Id;
	code: VarInt;
}

export interface StopSending {
	type: "stop_sending";
	id: Stream.Id;
	code: VarInt;
}

export interface ConnectionClose {
	type: "connection_close";
	code: VarInt;
	reason: string;
}

export interface Padding {
	type: "padding";
}

export interface Ping {
	type: "ping";
}

export type Any = Data | ResetStream | StopSending | ConnectionClose;

export function encode(frame: Any, version: Version = "webtransport"): Uint8Array {
	if (version === "webtransport") {
		return encodeWebTransport(frame);
	}
	return encodeQMux(frame);
}

function encodeWebTransport(frame: Any): Uint8Array {
	switch (frame.type) {
		case "stream": {
			let buffer = new Uint8Array(new ArrayBuffer(1 + 8 + frame.data.length), 0, 1);

			buffer[0] = frame.fin ? 0x09 : 0x08;
			buffer = frame.id.value.encode(buffer);

			buffer = new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength + frame.data.length);
			buffer.set(frame.data, buffer.byteLength - frame.data.length);

			return buffer;
		}

		case "reset_stream": {
			let buffer = new Uint8Array(new ArrayBuffer(1 + 8 + 8), 0, 1);

			buffer[0] = 0x04;
			buffer = frame.id.value.encode(buffer);
			buffer = frame.code.encode(buffer);
			return buffer;
		}

		case "stop_sending": {
			let buffer = new Uint8Array(new ArrayBuffer(1 + 8 + 8), 0, 1);

			buffer[0] = 0x05;
			buffer = frame.id.value.encode(buffer);
			buffer = frame.code.encode(buffer);
			return buffer;
		}

		case "connection_close": {
			const body = new TextEncoder().encode(frame.reason);
			let buffer = new Uint8Array(new ArrayBuffer(1 + 8 + body.length), 0, 1);

			buffer[0] = 0x1d;
			buffer = frame.code.encode(buffer);

			buffer = new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength + body.length);
			buffer.set(body, buffer.byteLength - body.length);

			return buffer;
		}
	}
}

function encodeQMux(frame: Any): Uint8Array {
	switch (frame.type) {
		case "stream": {
			// Always set LEN bit (0x02), type = 0x0a | fin_bit
			const frameType = VarInt.from(0x0a | (frame.fin ? 0x01 : 0x00));
			const lengthVi = VarInt.from(frame.data.length);

			const maxSize = 8 + 8 + 8 + frame.data.length;
			let buffer = new Uint8Array(new ArrayBuffer(maxSize), 0, 0);

			buffer = frameType.encode(buffer);
			buffer = frame.id.value.encode(buffer);
			buffer = lengthVi.encode(buffer);

			buffer = new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength + frame.data.length);
			buffer.set(frame.data, buffer.byteLength - frame.data.length);

			return buffer;
		}

		case "reset_stream": {
			const frameType = VarInt.from(0x04);
			const finalSize = VarInt.from(0);

			let buffer = new Uint8Array(new ArrayBuffer(8 + 8 + 8 + 8), 0, 0);

			buffer = frameType.encode(buffer);
			buffer = frame.id.value.encode(buffer);
			buffer = frame.code.encode(buffer);
			buffer = finalSize.encode(buffer);
			return buffer;
		}

		case "stop_sending": {
			const frameType = VarInt.from(0x05);

			let buffer = new Uint8Array(new ArrayBuffer(8 + 8 + 8), 0, 0);

			buffer = frameType.encode(buffer);
			buffer = frame.id.value.encode(buffer);
			buffer = frame.code.encode(buffer);
			return buffer;
		}

		case "connection_close": {
			// APPLICATION_CLOSE (0x1d)
			const frameType = VarInt.from(0x1d);
			const causingFrameType = VarInt.from(0);
			const body = new TextEncoder().encode(frame.reason);
			const reasonLength = VarInt.from(body.length);

			let buffer = new Uint8Array(new ArrayBuffer(8 + 8 + 8 + 8 + body.length), 0, 0);

			buffer = frameType.encode(buffer);
			buffer = frame.code.encode(buffer);
			buffer = causingFrameType.encode(buffer);
			buffer = reasonLength.encode(buffer);

			buffer = new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength + body.length);
			buffer.set(body, buffer.byteLength - body.length);

			return buffer;
		}
	}
}

export function decode(buffer: Uint8Array, version: Version = "webtransport"): Any | null {
	if (buffer.length === 0) {
		throw new Error("Invalid frame: empty buffer");
	}

	if (version === "webtransport") {
		return decodeWebTransport(buffer);
	}
	return decodeQMux(buffer);
}

function decodeWebTransport(buffer: Uint8Array): Any {
	const frameType = buffer[0];
	buffer = buffer.slice(1);

	let v: VarInt;

	if (frameType === 0x04) {
		[v, buffer] = VarInt.decode(buffer);
		const id = new Stream.Id(v);

		[v, buffer] = VarInt.decode(buffer);
		const code = v;

		return { type: "reset_stream", id, code };
	}

	if (frameType === 0x05) {
		[v, buffer] = VarInt.decode(buffer);
		const id = new Stream.Id(v);

		[v, buffer] = VarInt.decode(buffer);
		const code = v;

		return { type: "stop_sending", id, code };
	}

	if (frameType === 0x1d) {
		[v, buffer] = VarInt.decode(buffer);
		const code = v;

		const reason = new TextDecoder().decode(buffer);

		return { type: "connection_close", code, reason };
	}

	if (frameType === 0x08 || frameType === 0x09) {
		[v, buffer] = VarInt.decode(buffer);
		const id = new Stream.Id(v);

		return {
			type: "stream",
			id,
			data: buffer,
			fin: frameType === 0x09,
		};
	}

	throw new Error(`Invalid frame type: ${frameType}`);
}

function decodeQMux(buffer: Uint8Array): Any | null {
	let v: VarInt;

	[v, buffer] = VarInt.decode(buffer);
	const frameType = v.value;

	// STREAM frames: 0x08-0x0f
	if (frameType >= 0x08n && frameType <= 0x0fn) {
		const hasOff = (frameType & 0x04n) !== 0n;
		const hasLen = (frameType & 0x02n) !== 0n;
		const hasFin = (frameType & 0x01n) !== 0n;

		[v, buffer] = VarInt.decode(buffer);
		const id = new Stream.Id(v);

		// Skip offset if present
		if (hasOff) {
			[v, buffer] = VarInt.decode(buffer);
		}

		let data: Uint8Array;
		if (hasLen) {
			[v, buffer] = VarInt.decode(buffer);
			const len = Number(v.value);
			data = buffer.slice(0, len);
			buffer = buffer.slice(len);
		} else {
			data = buffer;
		}

		return { type: "stream", id, data, fin: hasFin };
	}

	// RESET_STREAM
	if (frameType === 0x04n) {
		[v, buffer] = VarInt.decode(buffer);
		const id = new Stream.Id(v);

		[v, buffer] = VarInt.decode(buffer);
		const code = v;

		// Skip final_size
		[v, buffer] = VarInt.decode(buffer);

		return { type: "reset_stream", id, code };
	}

	// STOP_SENDING
	if (frameType === 0x05n) {
		[v, buffer] = VarInt.decode(buffer);
		const id = new Stream.Id(v);

		[v, buffer] = VarInt.decode(buffer);
		const code = v;

		return { type: "stop_sending", id, code };
	}

	// CONNECTION_CLOSE / APPLICATION_CLOSE
	if (frameType === 0x1cn || frameType === 0x1dn) {
		[v, buffer] = VarInt.decode(buffer);
		const code = v;

		// Skip frame_type field
		[v, buffer] = VarInt.decode(buffer);

		// reason_length + reason
		[v, buffer] = VarInt.decode(buffer);
		const reasonLen = Number(v.value);
		const reason = new TextDecoder().decode(buffer.slice(0, reasonLen));

		return { type: "connection_close", code, reason };
	}

	// Flow control and other frames — ignore
	// MAX_DATA, MAX_STREAM_DATA, MAX_STREAMS, DATA_BLOCKED, etc.
	return null;
}
